src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/optimization/PersistenceOptimizer.java

```diff
@@ -0,0 +1,208 @@
+                }
+            }
+        }
+
+        //return pairs sorted by descending persistence (explicitly-typed lambda required here for use of reversed() to compile)
+        return extremaPairs.stream().sorted(Comparator.comparingDouble((ExtremaPair p) -> p.persistence).reversed()).collect(Collectors.toList());
+    }
+
+    //create a new component and add a new color when a local minimum is found
+    private static void createComponent(final double[] data,
+                                        final List<Component> components,
+                                        final int[] colors,
+                                        final int index) {
+        colors[index] = components.size();
+        components.add(new Component(index, data[index]));
+    }
+
+    //given components and colors, extend the component at componentIndex to include the point at datumIndex
+    //and assign the color of that component to this point
+    private static void extendComponent(final List<Component> components,
+                                        final int[] colors,
+                                        final int componentIndex,
+                                        final int datumIndex) {
+        if (datumIndex + 1 == components.get(componentIndex).leftIndex) {
+            components.get(componentIndex).leftIndex = datumIndex;
+        } else if (datumIndex - 1 == components.get(componentIndex).rightIndex) {
+            components.get(componentIndex).rightIndex = datumIndex;
+        }
+        colors[datumIndex] = componentIndex;
+    }
+
+    //merge components by assigning the color of the component with the lower minimum value
+    //to the boundary of the merged component
+    private static void mergeComponents(final List<Component> components,
+                                        final int[] colors,
+                                        final int index1,
+                                        final int index2) {
+        final int indexToKeep;
+        final int indexToMerge;
+        if (components.get(index1).minValue < components.get(index2).minValue) {
+            indexToKeep = index1;
+            indexToMerge = index2;
+        } else if (components.get(index1).minValue > components.get(index2).minValue) {
+            indexToKeep = index2;
+            indexToMerge = index1;
+        } else if (index1 < index2) {       //if values are equal, sort by index
+            indexToKeep = index1;
+            indexToMerge = index2;
+        } else {
+            indexToKeep = index2;
+            indexToMerge = index1;
+        }
+        colors[components.get(indexToMerge).leftIndex] = indexToKeep;
+        colors[components.get(indexToMerge).rightIndex] = indexToKeep;
+        if (components.get(indexToKeep).minIndex > components.get(indexToMerge).minIndex) {
+            components.get(indexToKeep).leftIndex = components.get(indexToMerge).leftIndex;
```

Comment:
Not removing the `components.get(indexToMerge)` component may introduce some space overhead, which could be linear in most degenerate examples, but I don't see a quick work-around..

Reply:
Technically true, but space considerations are probably not an issue in most cases (and I think actually removing "merged" components also complicates the bookkeeping a bit).